datafusion/sql/src/planner.rs

```diff
@@ -784,6 +784,9 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
 
                 let filter_expr = self.sql_to_rex(predicate_expr, &join_schema, ctes)?;
 
+                // reduce outer joins
+                let plans = reduce_outer_joins(plans, &filter_expr)?;
```

Comment:
This seems like it should be implemented as an optimization rule rather than in the SQL query planner and this would mean that we would get the same optimizations regardless of whether the query was created from SQL or from the  DataFrame API.